src/reconcile/rec_visibility.c

```diff
@@ -586,40 +587,64 @@ __wt_rec_upd_select(WT_SESSION_IMPL *session, WT_RECONCILE *r, WT_INSERT *ins, v
-         * nothing needs to be written.
+         * If the newest is a tombstone then select the update before it and set the end of the
+         * visibility window to its time pair as appropriate to indicate that we should return "not
+         * found" for reads after this point.
+         *
+         * Otherwise, leave the end of the visibility window at the maximum possible value to
+         * indicate that the value is visible to any timestamp/transaction id ahead of it.
          */
-        if ((upd_select->stop_ts == WT_TS_MAX && upd_select->stop_txn == WT_TXN_MAX) &&
-          ((upd_select->start_ts == WT_TS_NONE && upd_select->start_txn == WT_TXN_NONE) ||
-              __wt_txn_visible_all(session, upd_select->start_txn, upd_select->start_ts))) {
-            upd_select->start_ts = WT_TS_NONE;
-            upd_select->start_txn = WT_TXN_NONE;
-            upd_select->stop_ts = WT_TS_MAX;
-            upd_select->stop_txn = WT_TXN_MAX;
+        WT_ASSERT(session, upd_select->upd != NULL);
+        if (upd_select->upd->type == WT_UPDATE_TOMBSTONE) {
+            upd_select->stop_ts = upd_select->upd->start_ts;
+            upd_select->stop_txn = upd_select->upd->txnid;
+            upd_select->upd = upd_select->upd->next;
+        }
+        if (upd_select->upd != NULL) {
+            /* The beginning of the validity window is the selected update's time pair. */
+            upd_select->durable_ts = upd_select->start_ts = upd_select->upd->start_ts;
+            upd_select->start_txn = upd_select->upd->txnid;
+        } else {
+            /*
+             * The only time we don't get the unpacked value is when we insert a key in which case
+             * we can't have a tombstone.
+             */
+            WT_ASSERT(session, vpack != NULL);
+            /*
+             * It's possible to have a tombstone as the only update in the update list. If we
+             * reconciled before with only a single update and then read the page back into cache,
+             * we'll have an empty update list. And applying a delete on top of that will result in
+             * ONLY a tombstone in the update list.
+             *
+             * In this case, we should leave the selected update unset to indicate that we want to
+             * keep the same on-disk value but set the stop time pair to indicate that the validity
+             * window ends when this tombstone started.
+             */
+            upd_select->durable_ts = upd_select->start_ts = vpack->start_ts;
+            upd_select->start_txn = vpack->start_txn;
+            /*
+             * Leaving the update unset means that we can skip reconciling. If we've set the stop
+             * time pair because of a tombstone after the on-disk value, we still have work to do so
+             * that is NOT ok. Let's allocate an update equivalent to the on-disk value and continue
+             * on our way!
+             */
+            WT_RET(__wt_scr_alloc(session, 0, &tmp));
```

Comment:
This can't be `WT_RET`, we've already allocated `key`.

Come to think of it, there are `WT_ERR` calls before this line in this function, the dist scripts should have caught this case.

Reply:
3ebd6ac

Reply:
3ebd6ac

Reply:
Do we know why the dist script didn't catch this?

Reply:
It did catch it. I thought it was incorrectly flagging a different line and "fixed" the `dist` script (which I've reverted in the commit).